### PR TITLE
modify rowStartInfo for hiopMatrixRajaSparseTriplet

### PR DESCRIPTION
**Pull request by nychiang**
This PR addresses the following issues:
1. Compute `row_start` on CPU and save it for later use.
2. Add functions to synchronize the host and device data in RowStartsInfo.
3. Remove RAJA:inclusive_scan, and build row_start on CPU when FR is activated

### CODE REVIEW

**Review by cnpetra — APPROVED** [review]

File: src/LinAlg/hiopMatrixRajaSparseTriplet.cpp [hunk 1]
```diff
@@ -1074,17 +1099,17 @@ void hiopMatrixRajaSparseTriplet::set_Jac_FR(const hiopMatrixSparse& Jac_c,
...
 
-  if(J_d.row_starts_host == nullptr) {
-    J_d.row_starts_host = J_d.allocAndBuildRowStarts();
+  if(J_d.row_starts_ == nullptr) {
+    J_d.row_starts_ = J_d.allocAndBuildRowStarts();
   }
-  assert(J_d.row_starts_host);
-  int* Jd_row_st = J_d.row_starts_host->idx_start_;
+  assert(J_d.row_starts_);
+  int* Jd_row_st = J_d.row_starts_->idx_start_;
```
Comment (pelesh) [inline src/LinAlg/hiopMatrixRajaSparseTriplet.cpp:1113]: Should we use HiOp's `index_type` here and elsewhere in the code when accessing vector/matrix indices?

**Comment by pelesh** [comment]
Why are `idx_start_` and `idx_start_host_` encapsulated in the nested class `RowStartInfo`? It seems to me it would be better to have member variables of the sparse matrix as:
```c++
  int* iRow_;       ///< row indices of the nonzero entries
  int* iRowStarts_; ///< indices pointing to starts of new rows in values_
  int* jCol_;       ///< column indices of the nonzero entries
  double* values_;  ///< values of the nonzero entries
```
That way, it seems to me, it would be easier to set row start indices instead of going through a nested class. Also, this would allow you to instantiate CSR or COO matrix as you wish. This may be beyond the scope of this PR, though.

**Comment by pelesh** [comment]
The code builds smoothly and all tests pass.

However, when built in developer mode where additional compiler warning flags are added, the build produces a massive amount of warnings. Most of them seem easy to fix, but there may me some hidden in there that are pointing to potential issues. Also, CMake reported a nonfatal errors when developer mode was enabled.

CC @ashermancinelli

**Review by pelesh — CHANGES_REQUESTED** [review]
I suggest reviewing warnings in developer mode that pertain to this PR. Since these are only 3 files to check, it should not be too difficult. Some warnings may be pointing to actual bugs.

**Comment by nychiang** [comment]
> Why are `idx_start_` and `idx_start_host_` encapsulated in the nested class `RowStartInfo`? It seems to me it would be better to have member variables of the sparse matrix as:
> 
> ```c++
>   int* iRow_;       ///< row indices of the nonzero entries
>   int* iRowStarts_; ///< indices pointing to starts of new rows in values_
>   int* jCol_;       ///< column indices of the nonzero entries
>   double* values_;  ///< values of the nonzero entries
> ```
> 
> That way, it seems to me, it would be easier to set row start indices instead of going through a nested class. Also, this would allow you to instantiate CSR or COO matrix as you wish. This may be beyond the scope of this PR, though.

@pelesh Oh, I thought you introduced the nested class `RowStartInfo` and I just added the host buffer into your class. For sure I can make `irow_start_` as a member variable of the sparse matrix, but this is is a question about code design and I prefer getting an approval from the author of this code.

**Comment by nychiang** [comment]
@ashermancinelli @pelesh 
I received the same cmake error in the `develop` branch:
CMake Error: install(EXPORT "hiop-targets" ...) includes target "hiop_tpl" which requires target "hiop_warnings" that is not in any export set. I think we introduced this error in PR #279. Better to file another new PR to correct it.  

About the compiler warnings, I noticed that this PR doesn't introduce new warnings. 
We have 14944 warnings even in the develop branch.
In addition, I found that 14567 warnings come from unused parameters.
I think we should file a new PR to go through all these warnings.

**Comment by cnpetra** [comment]
> > Why are `idx_start_` and `idx_start_host_` encapsulated in the nested class `RowStartInfo`? It seems to me it would be better to have member variables of the sparse matrix as:
> > ```c++
> >   int* iRow_;       ///< row indices of the nonzero entries
> >   int* iRowStarts_; ///< indices pointing to starts of new rows in values_
> >   int* jCol_;       ///< column indices of the nonzero entries
> >   double* values_;  ///< values of the nonzero entries
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That way, it seems to me, it would be easier to set row start indices instead of going through a nested class. Also, this would allow you to instantiate CSR or COO matrix as you wish. This may be beyond the scope of this PR, though.
> 
> @pelesh Oh, I thought you introduced the nested class `RowStartInfo` and I just added the host buffer into your class. For sure I can make `irow_start_` as a member variable of the sparse matrix, but this is is a question about code design and I prefer getting an approval from the author of this code.

I authored the classes. The rationale for using a nested class `RowStartInfo` is that in most cases the functionality of  `RowStartInfo` is not needed and, thus,  the inner class and it's many members will not be instantiated. Not sure if this is still the case. I would let it as it is since will not help with anything at this point, performance or functionality wise.

File: src/LinAlg/hiopMatrixRajaSparseTriplet.cpp (hunk 1 above)
Comment (nychiang) [reply src/LinAlg/hiopMatrixRajaSparseTriplet.cpp]: Now I use `index_type` and `size_type` in class RowStartsInfo and the places to access these member variables.

**Pull request merged**